src/rgw/rgw_gc.cc

```diff
@@ -38,15 +38,20 @@ void RGWGC::initialize(CephContext *_cct, RGWRados *_store) {
     char buf[32];
     snprintf(buf, 32, ".%d", i);
     obj_names[i].append(buf);
-  }
 
-  // Create GC queues
-  for (int i = 0; i < max_objs; i++) {
-    ldpp_dout(this, 20) << "RGWGC::initialize initing gc queue with name = " << obj_names[i] << dendl;
+    auto it = transitioned_objects_cache.begin() + i;
+    transitioned_objects_cache.insert(it, false);
+
+    //version = 0 -> not ready for transition
+    //version = 1 -> marked ready for transition
     librados::ObjectWriteOperation op;
-    op.create(true);
-    uint64_t queue_size = 1048576, num_urgent_data_entries = 50;
+    obj_version objv;
+    cls_version_check(op, objv, VER_COND_EQ);
+    op.create(false);
+    const uint64_t queue_size = cct->_conf->rgw_gc_max_queue_size, num_urgent_data_entries = cct->_conf->rgw_gc_num_urgent_data_entries;
     cls_rgw_gc_queue_init(op, obj_names[i], queue_size, num_urgent_data_entries);
+    objv.ver = 1;
+    cls_version_set(op, objv);
     store->gc_operate(obj_names[i], &op);
```

Comment:
@cbodley : Should we handle errors other that -ECANCELED, -EPERM and -EEXISTS, and set version to 1, even if init queue returns an error?

Reply:
since this cls_version_set() comes after the cls_rgw_gc_queue_init() call in this transaction, it won't run unless cls_rgw_gc_queue_init() succeeds. and i think that's exactly what we want here :+1:

Reply:
yes, but what if init queue fails for reasons other than version not being right, or rgw gc module not being loaded (like decode error or write error in init queue), should we allow the entries to still go to omap?

Reply:
if the cls_rgw_gc queue init fails, it's not safe to write entries to it, so we do need to keep using omap as long as its cls_version is 0

Reply:
ok, so omap is going to serve as failback in case anything happens to the queue.